src/traces/heatmap/has_columns.js

```diff
@@ -0,0 +1,14 @@
+/**
+* Copyright 2012-2015, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+'use strict';
+
+module.exports = function(trace) {
+    return !Array.isArray(trace.z[0]);
```

Comment:
`hasColumns` is used in both the defaults and the calc ; gotta make a separate module to keep things :palm_tree: